client/gui/InterfaceBuilder.cpp

```diff
@@ -0,0 +1,208 @@
+/*
+* InterfaceBuilder.cpp, part of VCMI engine
+*
+* Authors: listed in file AUTHORS in main folder
+*
+* License: GNU General Public License v2.0 or later
+* Full text of license available in license.txt file, in main folder
+*
+*/
+
+#include "StdInc.h"
+
+#include "InterfaceBuilder.h"
+
+#include "../CGameInfo.h"
+#include "../gui/CAnimation.h"
+#include "../gui/CGuiHandler.h"
+#include "../widgets/CComponent.h"
+#include "../widgets/Buttons.h"
+#include "../widgets/MiscWidgets.h"
+#include "../widgets/ObjectLists.h"
+#include "../widgets/TextControls.h"
+#include "../windows/GUIClasses.h"
+#include "../windows/InfoWindows.h"
+
+#include "../../lib/CGeneralTextHandler.h"
+
+
+InterfaceBuilder::InterfaceBuilder(const JsonNode & config):
+	CIntObject()
+{
+	init(config);
+}
+
+InterfaceBuilder::InterfaceBuilder():
+	CIntObject()
+{
+}
+
+void InterfaceBuilder::addCallback(const std::string & callbackName, std::function<void(int)> callback)
+{
+	callbacks[callbackName] = callback;
+}
+
+const std::shared_ptr<CIntObject> InterfaceBuilder::widget(const std::string & name) const
+{
+	return widgets.at(name);
+}
+
+void InterfaceBuilder::init(const JsonNode &config)
+{
+	OBJ_CONSTRUCTION;
+	int unnamedObjectId = 0;
+	const std::string unnamedObjectPrefix = "__widget_";
+	
+	for(const auto & item : config["items"].Vector())
+	{
+		std::string name = item["name"].isNull()
+						? unnamedObjectPrefix + std::to_string(unnamedObjectId++)
+						: item["name"].String();
+		widgets[name] = buildWidget(item);
+	}
+}
+
+std::string InterfaceBuilder::buildText(const JsonNode & config) const
+{
+	if(config.isNull())
+		return "";
+	
+	if(config.isNumber())
+	{
+		return CGI->generaltexth->allTexts[config.Integer()];
+	}
+	return config.String();
+}
+
+std::shared_ptr<CIntObject> InterfaceBuilder::buildWidget(const JsonNode & config)
+{
+	assert(!config.isNull());
+	auto type = config["type"].String();
+	
+	int x = 0, y = 0;
+	if(!config["position"].isNull())
+	{
+		x = config["position"]["x"].Integer();
+		y = config["position"]["y"].Integer();
+	}
+	
+	std::string image, text = buildText(config["text"]);
```

Comment:
Not sure what's current coding style is, but I'd rather avoid declaring multiple variables in a single line.
Especially when you're initializing only one of them

Reply:
agree, looks weird